File: app/src/main/java/de/kuschku/libquassel/primitives/serializers/StringSerializer.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.primitives.serializers;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;
import java.nio.charset.Charset;

public class StringSerializer implements PrimitiveSerializer<String> {
    @NonNull
    private static final StringSerializer serializer = new StringSerializer();

    private StringSerializer() {
    }

    @NonNull
    public static StringSerializer get() {
        return serializer;
    }

    @Override
    public void serialize(@NonNull final ByteChannel channel, @Nullable final String data) throws IOException {
        if (data == null) {
            IntSerializer.get().serialize(channel, 0xffffffff);
        } else {
            final ByteBuffer contentBuffer = Charset.forName("UTF-16BE").encode(data);
            IntSerializer.get().serialize(channel, contentBuffer.limit());
            channel.write(contentBuffer);
        }
    }

    @Nullable
    @Override
    public String deserialize(@NonNull final ByteBuffer buffer) throws IOException {
        final int len = IntSerializer.get().deserialize(buffer);
        if (len == 0xffffffff)
            return null;
        else {
            final ByteBuffer contentBuffer = ByteBuffer.allocate(len);
            contentBuffer.put(buffer.array(), buffer.position(), len);
            contentBuffer.position(0);
            buffer.position(buffer.position() + len);
            return Charset.forName("UTF-16BE").decode(contentBuffer).toString();
        }
    }
}

File: lib/src/main/java/de/kuschku/libquassel/protocol/primitive/serializer/StringSerializer.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2019 Janne Koschinski
 * Copyright (c) 2019 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
@file:Suppress("NOTHING_TO_INLINE")
package de.kuschku.libquassel.protocol.primitive.serializer

import de.kuschku.libquassel.quassel.QuasselFeatures
import de.kuschku.libquassel.util.helpers.hexDump
import de.kuschku.libquassel.util.nio.ChainedByteBuffer
import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.Charset
import java.nio.charset.CharsetDecoder
import java.nio.charset.CharsetEncoder
import kotlin.concurrent.getOrSet

abstract class StringSerializer(
  private val charset: Charset,
  private val trailingNullBytes: Int
) : Serializer<String?> {
  constructor(charset: Charset, trailingNullByte: Boolean = false) :
    this(charset, if (trailingNullByte) 1 else 0)

  private val charBuffer = ThreadLocal<CharBuffer>()
  private val encoder = ThreadLocal<CharsetEncoder>()
  private val decoder = ThreadLocal<CharsetDecoder>()

  object UTF16 : StringSerializer(Charsets.UTF_16BE)
  object UTF8 : StringSerializer(Charsets.UTF_8)
  object C : StringSerializer(Charsets.ISO_8859_1, trailingNullByte = true)

  private inline fun charBuffer(len: Int): CharBuffer {
    val charBuffer = charBuffer.getOrSet {
      CharBuffer.allocate(1024)
    }

    val buf = if (len >= 1024)
      CharBuffer.allocate(len)
    else
      charBuffer
    buf.clear()
    buf.limit(len)
    return buf
  }

  private inline fun encoder() = encoder.getOrSet(charset::newEncoder)
  private inline fun decoder() = decoder.getOrSet(charset::newDecoder)

  override fun serialize(buffer: ChainedByteBuffer, data: String?, features: QuasselFeatures) =
    try {
      if (data == null) {
        IntSerializer.serialize(buffer, -1, features)
      } else {
        val charBuffer = charBuffer(data.length)
        charBuffer.put(data)
        charBuffer.flip()
        val encoder = encoder()
        encoder.reset()
        val byteBuffer = encoder.encode(charBuffer)
        IntSerializer.serialize(buffer, byteBuffer.remaining() + trailingNullBytes, features)
        buffer.put(byteBuffer)
        for (i in 0 until trailingNullBytes)
          buffer.put(0)
      }
    } catch (e: Throwable) {
      throw RuntimeException(data, e)
    }

  fun serialize(data: String?): ByteBuffer = try {
    if (data == null) {
      ByteBuffer.allocate(0)
    } else {
      val charBuffer = charBuffer(data.length)
      charBuffer.put(data)
      charBuffer.flip()
      val encoder = encoder()
      encoder.reset()
      encoder.encode(charBuffer)
    }
  } catch (e: Throwable) {
    throw RuntimeException(data, e)
  }

  fun deserializeAll(buffer: ByteBuffer): String? = try {
    val len = buffer.remaining()
    if (len == -1) {
      null
    } else {
      val limit = buffer.limit()
      buffer.limit(buffer.position() + len - trailingNullBytes)
      val charBuffer = charBuffer(len)
      val decoder = decoder()
      decoder.reset()
      decoder.decode(buffer, charBuffer, true)
      buffer.limit(limit)
      buffer.position(buffer.position() + trailingNullBytes)
      charBuffer.flip()
      charBuffer.toString()
    }
  } catch (e: Throwable) {
    buffer.hexDump()
    throw RuntimeException(e)
  }

  override fun deserialize(buffer: ByteBuffer, features: QuasselFeatures): String? = try {
    val len = IntSerializer.deserialize(buffer, features)
    if (len == -1) {
      null
    } else {
      val limit = buffer.limit()
      buffer.limit(buffer.position() + Math.max(0, len - trailingNullBytes))
      val charBuffer = charBuffer(len)
      val decoder = decoder()
      decoder.decode(buffer, charBuffer, true)
      buffer.limit(limit)
      buffer.position(buffer.position() + trailingNullBytes)
      charBuffer.flip()
      charBuffer.toString()
    }
  } catch (e: Throwable) {
    buffer.hexDump()
    throw RuntimeException(e)
  }
}
